libevmasm/RuleList.h

```diff
@@ -360,6 +376,30 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 		false
 	});
 
+	rules.push_back({
+		// AND(A, SHL(B, X)) -> SHL(B, X)
+		{Instruction::AND, {A, {Instruction::SHL, {B, X}}}},
+		[=]() -> Pattern { return {Instruction::SHL, {B, X}}; },
+		false,
+		[=]() {
+			if (B.d() > 256)
+				return false;
+			return boost::multiprecision::pow(u256(2), 256 - static_cast<int>(B.d())) - 1 == A.d();
```

Comment:
Hm, I think this is a little hard to understand. Might be easier with shifts because then it is also closer to the code we want to optimize?

Reply:
I don't understand what you mean by "might be easier with shifts"

Reply:
oh, you mean instead of using `pow`, using shifting.